lib/use-render-png.tsx

```diff
@@ -54,59 +54,36 @@ export function useRenderPNG({
     return () => abortContoller.abort("cleanup");
   }, []);
 
-  const downloadPNGRef = useRef(
-    () => new Promise((resolve, reject) => reject("not ready")),
-  );
-  const copyPNGRef = useRef(
-    () => new Promise((resolve, reject) => reject("not ready")),
-  );
+  const onDownload = (type: Exclude<RenderStatus, "idle">) => async () => {
+    if (!fontsRef.current) {
+      throw new Error("fonts not loaded");
+    }
 
-  useEffect(() => {
-    copyPNGRef.current = async () => {
-      if (!fontsRef.current) {
-        throw new Error("fonts not loaded");
-      }
+    setRenderStatus(type);
 
-      setRenderStatusRef.current("copying");
-
-      const { blob } = await renderPNG({
-        fonts: fontsRef.current,
-        videoDetails,
-        theme,
-      });
-
-      await navigator.clipboard.write([
-        new ClipboardItem({ "image/png": blob }),
-      ]);
-      toast.success("Image copié dans ton presse papier");
-
-      setRenderStatusRef.current("idle");
-    };
-    downloadPNGRef.current = async () => {
-      if (!fontsRef.current) {
-        throw new Error("fonts not loaded");
-      }
+    const { blob } = await renderPNG({
+      fonts: fontsRef.current,
+      videoDetails,
+      theme,
+    });
 
-      setRenderStatusRef.current("downloading");
+    if (type === "copying") {
+      await copyAction(blob);
+    } else {
+      await downloadAction(blob, videoDetails.title);
+    }
 
-      const { blob } = await renderPNG({
-        fonts: fontsRef.current,
-        videoDetails,
-        theme,
-      });
-      const url = URL.createObjectURL(blob);
+    setRenderStatus("idle");
 
-      const a = document.createElement("a");
-      a.href = url;
-      a.download = `${videoDetails.title}.png`;
-      a.click();
+    return blob;
+  };
 
-      setRenderStatusRef.current("idle");
-    };
-  }, [fontsRef, theme, videoDetails]);
+  const downloadPNG = onDownload("downloading");
 
-  const downloadPNG = useCallback(() => downloadPNGRef.current(), []);
-  const copyPNG = useCallback(() => copyPNGRef.current(), []);
+  const copyPNG = onDownload("copying");
 
-  return useMemo(() => ({ downloadPNG, copyPNG }), [downloadPNG, copyPNG]);
+  return {
```

Comment:
As you pointed in your comment, the problem with not memoizing the functions is that they're going to be recreated every time the parent component re-renders. But this hook has no control over how often its parent re-renders and how the function it creates are used.

For example, they could be used in a `useEffect` and passed as a dependency. Recreating them every time runs the risk of triggering `useEffect`s too often, among other things.

So that's why I prefer to memoize in custom hooks, because they are meant to be reused in different contexts you don't have control over.

On the other hand I would be ok not to memoize if the hook wasn't reusable and only used locally in a controlled context.

With that being said, there are a few ideas I like with your changes so I'll keep them 🙏

Reply:
Yes, I understand, but you should not use useEffect (troll) ❤️ Thanks, bro

Reply:
Done in e12d17be920b4002cb23bf6321db8aaa6b3f5c32, thanks again @Melvynx 🙏